### PR TITLE
DCON-4860 Never surface transient auth infra failures as 401 (INC-322)

### PR DESCRIPTION
**Pull request by imranq2**
## Summary
- Production incident INC-322: when Redis evicted cached JWKS tokens (memory exhaustion), the auth layer returned a hard 401 "grant removed" instead of a 503/fallback — signaling downstream systems (Google) to permanently drop grants, turning a temporary infrastructure hiccup into permanent auth failures for ~65,000 users.
- Threads a transient/infra-failure marker (`isTransient`/`statusCode = 503`) through the auth chain instead of swallowing errors into empty results:
  - `AuthService.getJwksByUrlAsync`/`getExternalJwksAsync`: rethrow a marked error on fetch failure instead of returning `{keys: []}`/`[]` — an empty keyset is indistinguishable downstream from "this endpoint legitimately has no keys" (permanent), when the truth is "we couldn't reach it right now" (transient). The existing `jwksFetchInFlight` in-flight-request coalescing is preserved; a failure is never cached as a valid empty response.
  - `AuthService.verify()`: userinfo-endpoint fetch failures now call `done(error)` (passport's real-error signature) instead of `done(null, false, {...})` (passport's fail signature, which maps to 401).
  - `jwt.bearer.strategy.js`: extracted `handleSigningKeyError` into a named, exported function. A genuine `SigningKeyNotFoundError` after a **successful** JWKS fetch (the token's `kid` really doesn't match any known key) is untouched — still a real, permanent 401. Any other error (network/timeout/DNS) is now marked transient/503.
  - `authentication.middleware.js`: `authenticateWithJsonFailure` now checks passport's failure `info` for the transient marker and responds 503 (mirroring the existing GraphQL-vs-REST route pattern) instead of unconditionally sending a 401.
- Un-quarantines `src/tests/unit/strategies/authFailureMode.test.js` (removed from `jest.config.js`'s `testPathIgnorePatterns`) now that all 8 tests pass. Also fixes two bugs uncovered in that same test file (a missing try/catch around an intentionally-throwing call, and a test that exercised a local hardcoded stub instead of the real `handleSigningKeyError`), and updates `src/tests/strategies/authService.test.js`/`src/tests/unit/strategies/authService.test.js`, whose pre-existing assertions encoded the old buggy swallow-to-empty/401 behavior.

## Test plan
- [x] `authFailureMode.test.js` — 8/8 passing
- [x] `src/tests/unit/strategies/`, `src/tests/strategies/`, `src/tests/unit/middleware/fhir/authentication.middleware.test.js`, `src/tests/unit/oauth/` — 8 suites / 163 tests passing, no regressions
- [x] `jwtCacheThunderingHerd.test.js` run standalone (still quarantined, entry untouched) — 6/6 passing, no regression
- [x] Lint clean on all changed files
- [ ] `src/tests/auth_failure/auth_failure_response.test.js` (Mongo-backed integration test) — ran with the fix applied, 8/8 passing on 2 of 3 attempts; one run flaked (404 vs expected 401), but confirmed via `git stash` that the **unfixed base commit flakes too** here (different failure mode, ECONNRESET) — this is sandbox-level Mongo/network flakiness in the dev environment, not caused by this change. Should be re-run in CI to confirm.
- [ ] Security review against `review.md` by someone else familiar with this surface, per this repo's `CLAUDE.md`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/strategies/jwt.bearer.strategy.js [hunk 1]
```diff
@@ -9,6 +9,37 @@ const {assertTypeEquals} = require("../utils/assertType");
...
+    }
+    logError('JWKS signing key error', {
+        user: '',
+        args: {error: err.message, type: err.name}
+    });
+    err.isTransient = true;
+    if (!err.statusCode) {
+        err.statusCode = 503;
+    }
+    return cb(err);
```
[the remark this answers is not in the document]
Comment (imranq2) [reply src/strategies/jwt.bearer.strategy.js]: Fixed in `9bc81958d`: `handleSigningKeyError` now builds a new error object (`Object.assign(Object.create(Object.getPrototypeOf(err)), err)`, with `message`/`stack` copied explicitly since they're non-enumerable on native `Error` instances) and sets `isTransient`/`statusCode` on that clone instead of mutating the caller-owned `err`.

File: src/strategies/authService.js [hunk 2]
```diff
@@ -182,7 +192,14 @@ class AuthService {
...
+                // an infrastructure outage look like "no external keys configured", which
+                // downstream turns into a permanent SigningKeyNotFoundError -> 401 (INC-322).
+                error.isTransient = true;
+                if (!error.statusCode) {
+                    error.statusCode = 503;
+                }
+                throw error;
             }
         }
         return [];
```
[the remark this answers is not in the document]
Comment (imranq2) [reply src/strategies/authService.js]: Fixed in `9bc81958d`: `getExternalJwksAsync` now uses `Promise.allSettled` over `extJwksUrls` instead of `async.map`, collects keys from every fulfilled result, logs the failures, and only throws the transient/503 error when every URL failed. Added test "partial JWKS provider failure still returns keys from healthy providers (INC-322)" (one down, one healthy -> keys from the healthy one are returned, failure is logged) and "throws a transient error when every configured JWKS provider fails (INC-322)". 153/153 relevant tests pass, lint clean.

[the remark this answers is not in the document]
Comment (imranq2) [reply src/strategies/authService.js]: Fixed in `9bc81958d`: `WellKnownConfigurationManager.getJwksUrlsAsync` now distinguishes "zero URLs configured" (returns `[]` immediately, unchanged) from "every configured URL failed to fetch" (throws an `isTransient`/`statusCode: 503` error) from "some succeeded" (returns the successful subset). `getExternalJwksAsync`'s well-known-fallback branch lets that propagate rather than falling through to the unguarded `return []`. Added test "propagates a transient error when every well-known URL fails to resolve JWKS URLs (INC-322)" plus coverage directly in `wellKnownConfigurationManager.test.js` for the three-way distinction. Verified passing.

File: src/middleware/fhir/authentication.middleware.js [hunk 3]
```diff
@@ -32,6 +62,19 @@ const authenticateWithJsonFailure = (strategy, options = {session: false}) => {
...
+                        info.statusCode = 503;
+                    }
+                    if (req.isGraphQLRoute) {
+                        return next(info);
+                    }
+                    return sendServiceUnavailableJson(res);
+                }
                 // Classify the failure for audit logging
                 if (info && info.message === 'No auth token') {
                     req.authFailureDetail = 'No token available';
```
[the remark this answers is not in the document]
Comment (imranq2) [reply src/middleware/fhir/authentication.middleware.js:114]: Fixed in `9bc81958d`: added a shared `respondToTransientAuthFailure(req, res, next)` helper that constructs a sanitized error (`new Error('Authentication service temporarily unavailable')`, `statusCode: 503`) for the GraphQL branch instead of forwarding raw `info`, and reused it for the REST 503 path. Also added a check for `isTransientAuthFailure(err)` at the top `if (err)` branch in `authenticateWithJsonFailure` — this covers the `AuthService.verify()` userinfo-endpoint-failure path, which reaches here via `done(error)`/`self.error()` (the `err` arg), not `self.fail()`/`info`, so it was bypassing the transient-failure handling entirely before. Both leak paths now route through the same sanitized response. Verified with new tests in `authentication.middleware.test.js`.

**Review by alvinhenrick — APPROVED** [review]
Approved. Strongest of the batch. Confirmed no fail-open: genuine invalid/expired/kid-mismatch tokens still return 401; only true infra failures (JWKS outage, userinfo error) become 503. All four prior findings resolved with tests. One follow-up: add a negative test pinning SigningKeyNotFoundError -> 401 so the transient/permanent boundary can't silently regress to fail-open. Rebase before merge.
